### PR TITLE
ROX-10613: Use ubi-minimal for scanner-db

### PR DESCRIPTION
**Pull request by janisz**
Refs: https://stackoverflow.com/a/68217042/1387612
Tests: https://github.com/openshift/release/pull/32858
OSCI: https://github.com/openshift/release/pull/32848

### CODE REVIEW

**Comment by janisz** [comment]
I guess to test it we need to first merge https://github.com/openshift/release/pull/32848

File: image/db/rhel/Dockerfile [hunk 1]
```diff
@@ -25,18 +25,20 @@ COPY --from=extracted_bundle /bundle/etc/postgresql.conf /bundle/etc/pg_hba.conf
...
 ARG POSTGRESQL_ARCH=x86_64
 
-RUN groupadd -g 70 postgres && \
+RUN microdnf install -y shadow-utils && \
+    groupadd -g 70 postgres && \
     adduser postgres -u 70 -g 70 -d /var/lib/postgresql -s /bin/sh && \
-    dnf upgrade -y && \
-    dnf install -y \
+    curl https://download.postgresql.org/pub/repos/yum/reporpms/EL-8-${POSTGRESQL_ARCH}/pgdg-redhat-repo-latest.noarch.rpm -o /tmp/pg_repo.rpm && \
+    rpm -i /tmp/pg_repo.rpm && \
```
Comment (RTann) [inline image/db/rhel/Dockerfile:32]: This step failed for me when I tried running this locally:

```
[root@beb09ad19088 /]# rpm -i /tmp/pg_repo.rpm
error: open of <html> failed: No such file or directory
error: open of <head><title>404 failed: No such file or directory
error: open of Not failed: No such file or directory
error: open of Found</title></head> failed: No such file or directory
error: open of <body> failed: No such file or directory
error: open of <center><h1>404 failed: No such file or directory
error: open of Not failed: No such file or directory
error: open of Found</h1></center> failed: No such file or directory
error: open of <hr><center>nginx</center> failed: No such file or directory
error: open of </body> failed: No such file or directory
error: open of </html> failed: No such file or directory
```

Comment (RTann) [reply image/db/rhel/Dockerfile:32]:
```
[root@beb09ad19088 /]# ls -l /tmp/pg_repo.rpm 
-rw-r--r-- 1 root root 146 Oct  4 22:02 /tmp/pg_repo.rpm
```

Comment (RTann) [reply image/db/rhel/Dockerfile:32]: I feel like we tried this before and ran into the same issue, but I cannot find any PRs related to this at the moment

Comment (janisz) [reply image/db/rhel/Dockerfile:32]: It's working for me
```
docker run -it registry.access.redhat.com/ubi8/ubi-minimal:8.6 /bin/bash
[root@a7f6fa9ee662 /]# export POSTGRESQL_ARCH=x86_64
[root@a7f6fa9ee662 /]# curl https://download.postgresql.org/pub/repos/yum/reporpms/EL-8-${POSTGRESQL_ARCH}/pgdg-redhat-repo-latest.noarch.rpm -o /tmp/pg_repo.rpm
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 13168  100 13168    0     0  24475      0 --:--:-- --:--:-- --:--:-- 24475
[root@a7f6fa9ee662 /]# rpm -i /tmp/pg_repo.rpm
warning: /tmp/pg_repo.rpm: Header V4 DSA/SHA1 Signature, key ID 442df0f8: NOKEY
[root@a7f6fa9ee662 /]# echo $?
0
```

Comment (janisz) [reply image/db/rhel/Dockerfile:32]: prow successfully build it https://quay.io/repository/stackrox-io/scanner-db/manifest/sha256:7da73ef2b03b0c4e713664c35c33f41317ff90763b60572a919409b3d84c097e?tab=layers

**Review by vladbologa — COMMENTED** [review]
I gave this a try, because I wanted to see whether local arm64 builds would still work with these changes.

And it seems that they don't. I'm running into a similar issue like here:
https://github.com/stackrox/stackrox/blob/989dbbaca060d22745d03e75f427e5841d0d3d5c/image/postgres/create-bundle.sh#L36

But I can't make the same workaround as @connorgorman did, because `microdnf` doesn't implement the `--nogpgcheck` option.

This is just FYI, I guess arm builds are not very important atm.

**Comment by janisz** [comment]
@vladbologa Thanks for testing. Could you try now?

File: image/db/rhel/Dockerfile [hunk 2]
```diff
@@ -25,18 +25,30 @@ COPY --from=extracted_bundle /bundle/etc/postgresql.conf /bundle/etc/pg_hba.conf
...
-RUN groupadd -g 70 postgres && \
+RUN microdnf install -y shadow-utils && \
+    groupadd -g 70 postgres && \
     adduser postgres -u 70 -g 70 -d /var/lib/postgresql -s /bin/sh && \
-    dnf upgrade -y && \
-    dnf install -y \
+    curl -sSLf https://download.postgresql.org/pub/repos/yum/RPM-GPG-KEY-PGDG-${PG_MAJOR} -o /tmp/pg_repo.key && \
+    rpm --import /tmp/pg_repo.key && \
+    curl -sSLf https://download.postgresql.org/pub/repos/yum/RPM-GPG-KEY-PGDG-AARCH64 -o /tmp/pg_repo.key && \
+    rpm --import /tmp/pg_repo.key && \
```
Comment (RTann) [inline image/db/rhel/Dockerfile:34]: Why do we need both of these?

File: image/db/rhel/Dockerfile (hunk 1 above)
Comment (RTann) [reply image/db/rhel/Dockerfile:36]: You're right, I missed the ARCH :(

File: image/db/rhel/Dockerfile (hunk 2 above)
Comment (RTann) [reply image/db/rhel/Dockerfile:34]: From the looks of it, I was able to succeed with just the first one:
```
[root@f9e6d3a223f7 /]# curl -sSLf https://download.postgresql.org/pub/repos/yum/RPM-GPG-KEY-PGDG-12 -o /tmp/pg_repo.key
[root@f9e6d3a223f7 /]# rpm --import /tmp/pg_repo.key
[root@f9e6d3a223f7 /]# curl -sSLf https://download.postgresql.org/pub/repos/yum/reporpms/EL-8-x86_64/pgdg-redhat-repo-latest.noarch.rpm -o /tmp/pg_repo.rpm
[root@f9e6d3a223f7 /]# rpm -i /tmp/pg_repo.rpm
[root@f9e6d3a223f7 /]# rpm -qa | grep pgdg
pgdg-redhat-repo-42.0-27.noarch
```

File: image/db/rhel/Dockerfile [hunk 3]
```diff
@@ -25,18 +25,30 @@ COPY --from=extracted_bundle /bundle/etc/postgresql.conf /bundle/etc/pg_hba.conf
...
-        https://download.postgresql.org/pub/repos/yum/reporpms/EL-8-${POSTGRESQL_ARCH}/pgdg-redhat-repo-latest.noarch.rpm && \
-    dnf install -y postgresql12-server && \
-    dnf clean all && \
+        postgresql12-server && \
+    # The removal of /usr/share/zoneinfo from UBI minimal images is intentional.
+    # After building the image, the image is reduced in size as much as possible,
+    # and the /usr/share/zoneinfo directory is purged as it saves space
+    # in the final distribution of the image.
+    # https://access.redhat.com/solutions/5616681
+    microdnf reinstall tzdata && \
```
Comment (RTann) [inline image/db/rhel/Dockerfile:46]: Do we need this?

File: image/db/rhel/Dockerfile [hunk 4]
```diff
@@ -25,18 +25,30 @@ COPY --from=extracted_bundle /bundle/etc/postgresql.conf /bundle/etc/pg_hba.conf
 
 ARG POSTGRESQL_ARCH=x86_64
 
-RUN groupadd -g 70 postgres && \
+RUN microdnf install -y shadow-utils && \
```
Comment (RTann) [inline image/db/rhel/Dockerfile:28]: should we run the `microdnf upgrade` before installing `shadow-utils`?

File: image/db/rhel/Dockerfile (hunk 2 above)
Comment (vladbologa) [reply image/db/rhel/Dockerfile:34]: The second one is for arm. But on my M1 Mac the GPG check was failing even with it.

File: image/db/rhel/Dockerfile (hunk 3 above)
Comment (janisz) [reply image/db/rhel/Dockerfile:46]: Yes, without it db will not start as it try to open `/usr/share/zoneinfo`

**Review by vladbologa — CHANGES_REQUESTED** [review]

**Review by vladbologa — APPROVED** [review]
works on arm64, thanks!

**Comment by RTann** [comment]
Let's keep the gpg key check for x86_64. We can keep it out for aarch64 since that's just for dev purposes (though let's also add a comment saying we are skipping it due to the fact it's just for dev purposes)

File: image/db/rhel/Dockerfile [hunk 5]
```diff
@@ -25,18 +25,27 @@ COPY --from=extracted_bundle /bundle/etc/postgresql.conf /bundle/etc/pg_hba.conf
 
 ARG POSTGRESQL_ARCH=x86_64
 
-RUN groupadd -g 70 postgres && \
+RUN curl -sSLf https://download.postgresql.org/pub/repos/yum/reporpms/EL-8-${POSTGRESQL_ARCH}/pgdg-redhat-repo-latest.noarch.rpm -o /tmp/pg_repo.rpm && \
```
[the remark this answers is not in the document]
Comment (janisz) [reply image/db/rhel/Dockerfile:28]: Previously we don't need that and it's in repo definition rpm so I think we don't need to add it manually. Something is broken with arm package I think it's build with different gpg key then used to sign.

**Comment by janisz** [comment]
✔️  Let's keep the gpg key check for x86_64.
✔️  let's also add a comment saying we are skipping it due to the fact it's just for dev purposes
@RTann PTAL

File: image/db/rhel/Dockerfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (RTann) [reply image/db/rhel/Dockerfile:28]: Can we re-add it here just for x86_64 prior to `rpm -i /tmp/pg_repo.rpm` to remove the warning that's generated?

**Review by RTann — APPROVED** [review]
Confirmed the change has been made as expected between both files:

```
$ diff -du image/db/rhel/Dockerfile image/db/rhel/Dockerfile.slim 
--- image/db/rhel/Dockerfile	2022-10-24 11:30:22.000000000 -0700
+++ image/db/rhel/Dockerfile.slim	2022-10-24 11:30:22.000000000 -0700
@@ -10,7 +10,7 @@
 
 FROM ${BASE_REGISTRY}/${BASE_IMAGE}:${BASE_TAG} AS base
 
-LABEL name="scanner-db" \
+LABEL name="scanner-db-slim" \
       vendor="StackRox" \
       maintainer="support@stackrox.com" \
       summary="Image scanner database for the StackRox Kubernetes Security Platform" \
@@ -57,7 +57,7 @@
 # This is equivalent to postgres:postgres.
 USER 70:70
 
-COPY --from=extracted_bundle /bundle/docker-entrypoint-initdb.d/definitions.sql.gz /docker-entrypoint-initdb.d/
+ENV ROX_SLIM_MODE="true"
 
 ENTRYPOINT ["docker-entrypoint.sh"]
```

**Comment by RTann** [comment]
@janisz it looks like you'll need to rebase for CI to pass

**Comment by janisz** [comment]
Are you sure. I think we need to merge osci release config change first and then CI will pass.

**Comment by RTann** [comment]
> Are you sure. I think we need to merge osci release config change first and then CI will pass.

that, too, but `ci/prow/images` shows me this as well:

```
error: Your local changes to the following files would be overwritten by checkout:
	.openshift-ci/build/Dockerfile.generate-genesis-dump
```

**Comment by janisz** [comment]
Rebased

**Pull request merged**